src/main/kotlin/app/revanced/patcher/apk/Apk.kt

```diff
@@ -0,0 +1,353 @@
+                    map.forEach { (name, value) -> getOrCreateEntry(name).setTo(value) }
+                }
+            }
+        }
+
+        /**
+         * Open a [app.revanced.patcher.apk.ResourceFile]
+         */
+        fun openFile(path: String) = ResourceFile(
+            resources.getHandle(path), archive, this
+        )
+
+        fun editXmlFile(path: String) = DomFileEditor(openFile(path))
+    }
+
+    val resources = Resources(module.tableBlock)
+
+    internal inner class BytecodeData {
+        private val dexFile = MultiDexContainerBackedDexFile(object : MultiDexContainer<DexBackedDexFile> {
+            /**
+             * Load all dex files from the [ApkModule] and create an entry for each of them.
+             */
+            private val entries = module.listDexFiles().map {
+                BasicDexEntry(
+                    this,
+                    it.name,
+                    it.openStream().use { stream -> RawDexIO.readRawDexFile(stream, it.length, null) })
+            }.associateBy { it.entryName }
+
+            override fun getDexEntryNames() = entries.keys.toList()
+            override fun getEntry(entryName: String) = entries[entryName]
+        })
+        private val opcodes = dexFile.opcodes
+
+        /**
+         * The classes and proxied classes of the [Base] apk file.
+         */
+        val classes = ProxyBackedClassList(dexFile.classes)
+
+        /**
+         * Write [classes] to the archive.
+         *
+         */
+        internal fun writeDexFiles() {
+            // Make sure to replace all classes with their proxy.
+            val classes = classes.also(ProxyBackedClassList::applyProxies)
+            val opcodes = opcodes
+
+            // Create patched dex files.
+            mutableMapOf<String, MemoryDataStore>().also {
+                val newDexFile = object : org.jf.dexlib2.iface.DexFile {
+                    override fun getClasses() = classes.toSet()
+                    override fun getOpcodes() = opcodes
+                }
+
+                // Write modified dex files.
+                MultiDexIO.writeDexFile(
+                    true, -1, // Core count.
+                    it, Patcher.dexFileNamer, newDexFile, DexIO.DEFAULT_MAX_DEX_POOL_SIZE, null
+                )
+            }.forEach { (name, store) ->
+                module.apkArchive.add(object : InputSource(name) {
+                    override fun getMethod() = ZipEntry.DEFLATED
+                    override fun getLength(): Long = store.size.toLong()
+                    override fun openStream() = store.readAt(0)
+                })
+            }
+        }
+    }
+
+    /**
+     * Metadata about an [Apk] file.
+     *
+     * @param packageName The package name of the [Apk] file.
+     * @param packageVersion The package version of the [Apk] file.
+     */
+    data class PackageMetadata(val packageName: String, val packageVersion: String) {
```

Comment:
Redundancy is not a problem. All of this are ref types and won't consume considerably more memory. I am not sure what you mean else, the constructor can be swapped so that you construct a PackageMetadata off the other type.

Reply:
It is a data class though. Thats the reason why it has a secondary constructor...